File: Graph/1042.Flower_Planting_With_No_Adjacent/Python/1042.Flower_Planting_With_No_Adjacent.py
class Solution(object):
    def gardenNoAdj(self, n, paths):
        """
        :type n: int
        :type paths: List[List[int]]
        :rtype: List[int]
        """
        #===============================#
        # Greedy-based traversal method #
        #===============================#

        ############
        #Initialize
        ##### Length of paths array #####
        len_paths = len(paths)

        ##### Root-child dictionary/hashmap #####
        root_child_dict = {}

        ##### Result array #####
        res_arry = [0 for _ in range(n)]


        ##############################################################
        #Greedy-based loop traversal with recorded dictionary/hashmap
        ##### Step 1: Record indexed-nodes information with dictionary/hashmap #####
        for paths_idx in range(len_paths):
            curr_node, next_node = ((paths[paths_idx])[0] - 1), ((paths[paths_idx])[1] - 1) #Current, Next node

            ##### Check if the current node existed or not #####
            if (curr_node not in root_child_dict):
                root_child_dict[curr_node] = [next_node] #Keep updating/recording
            else:
                (root_child_dict[curr_node]).append(next_node) #Keep updating/recording

            ##### Check if the next node existed or not #####
            if (next_node not in root_child_dict):
                root_child_dict[next_node] = [curr_node] #Keep updating/recording
            else:
                (root_child_dict[next_node]).append(curr_node) #Keep updating/recording
        
        ##### Step 2: Looped-traversal with recorded dictionary/hashmap #####
        for n_idx in range(n):
            color_arry = [False for _ in range(5)] #Color array

            ##### Check if the current node existed or not #####
            if (n_idx in root_child_dict):
                for next_node in root_child_dict[n_idx]:
                    color_arry[res_arry[next_node]] = True #Keep updating/recording
            else:
                pass

            for color_idx in range(4, 0, (-1)):

                ##### Check if the current indexed-color matched conditions or not #####
                if (not color_arry[color_idx]):
                    res_arry[n_idx] = color_idx #Keep updating/overwriting
                else:
                    pass

        return res_arry

File: Graph/1042.Flower_Planting_With_No_Adjacent/C++/1042.Flower_Planting_With_No_Adjacent.cpp
class Solution {
public:
    vector<int> gardenNoAdj(int n, vector<vector<int>> & paths) 
    {
        //=============================================//
        // Input type:                                 //
        //   - int (n-th nodes)                        //
        //   - vector<vector<int>> ref. (paths vector) //
        // Return type:                                //
        //   - vector<int> (result vector)             //
        //=============================================//

        /*Initialize*/
        ///// Length of paths vector /////
        int len_paths = paths.size();

        ///// Root-child dictionary, hashmap /////
        unordered_map<int, vector<int>> root_child_map;

        ///// Result vector /////
        vector<int> res_vect(n, 0);


        /*Greedy-based loop traversal with recorded dictionary, hashmap*/
        ///// Step 1: Record indexed-nodes information with dictionary, hashmap /////
        for (int paths_idx = 0; (paths_idx < len_paths); (paths_idx++)) //Whole
        {
            (root_child_map[((paths[paths_idx])[0] - 1)]).emplace_back(((paths[paths_idx])[1] - 1)); //Keep updating, recording
            (root_child_map[((paths[paths_idx])[1] - 1)]).emplace_back(((paths[paths_idx])[0] - 1)); //Keep updating, recording

        } //Whole

        ///// Step 2: Looped-traversal with recorded dictionary, hashmap /////
        for (int n_idx = 0; (n_idx < n); (n_idx++)) //N-th nodes
        {
            vector<bool> color_vect(5, false); //Color vector

            for (const int & next_node: root_child_map[n_idx]) { color_vect[res_vect[next_node]] = true; } //Keep updating, overwriting

            for (int color_idx = 4; (color_idx > 0); (color_idx--)) //Colors
            {
                ///// Check if the current indexed-color matched conditions or not /////
                if (! color_vect[color_idx]) { res_vect[n_idx] = color_idx; } //Keep updating, overwriting
                else { ; }

            } //Colors

        } //N-th nodes

        return res_vect;
    }
};
